heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java

```diff
@@ -104,12 +126,124 @@ public void updateTopology(final PackingPlans.PackingPlan existingProtoPackingPl
     // delete the physical plan so TMaster doesn't try to re-establish it on start-up.
     logFine("Deleted Physical Plan: %s", stateManager.deletePhysicalPlan(topologyName));
 
-    if (removableContainerCount > 0) {
+    // reactivate topology
+    reactivateTopology(stateManager, updatedTopology, removableContainerCount);
+
+    if (removableContainerCount > 0 && scalableScheduler.isPresent()) {
       scalableScheduler.get().removeContainers(containerDelta.getContainersToRemove());
     }
   }
 
   @VisibleForTesting
+  void deactivateTopology(SchedulerStateManagerAdaptor stateManager,
+                          final TopologyAPI.Topology topology)
+      throws InterruptedException {
+
+    List<TopologyAPI.Config.KeyValue> topologyConfig = topology.getTopologyConfig().getKvsList();
+    long deactivateSleepSeconds = TopologyUtils.getConfigWithDefault(
+        topologyConfig, TOPOLOGY_UPDATE_DEACTIVATE_WAIT_SECS, 0L);
+
+    logInfo("Deactivating topology %s before handling update request", topology.getName());
+    assertTrue(TMasterUtils.transitionTopologyState(
+            topology.getName(), TMasterUtils.TMasterCommand.DEACTIVATE, stateManager,
+            TopologyAPI.TopologyState.RUNNING, TopologyAPI.TopologyState.PAUSED),
+        "Failed to deactivate topology %s. Aborting update request", topology.getName());
+
+    if (deactivateSleepSeconds > 0) {
```

Comment:
Some typologies flush internal state to external systems every N seconds. In that case they would set this value to something lager than N. This assures that the topology deactivated (pausing new tuple consumption from spouts) and state is flushed before scaling occurs.